test/jdk/java/nio/channels/DatagramChannel/SRTest.java

```diff
@@ -24,184 +24,220 @@
+                    future.completeExceptionally(ex);
+                    return null;
+                }));
+        future.join();
     }
 
-    public static class ClassicWriter implements Sprintable {
+    public static class ClassicWriter implements Runnable, AutoCloseable {
+        final DatagramSocket ds;
         final int port;
-        Exception e = null;
 
-        ClassicWriter(int port) {
+        ClassicWriter(int port) throws SocketException {
             this.port = port;
-        }
-
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
+            this.ds = new DatagramSocket();
         }
 
         public void run() {
             try {
-                DatagramSocket ds = new DatagramSocket();
                 String dataString = "hello";
                 byte[] data = dataString.getBytes();
-                InetAddress address = InetAddress.getLocalHost();
+                InetAddress address = InetAddress.getLoopbackAddress();
                 DatagramPacket dp = new DatagramPacket(data, data.length,
                                                        address, port);
                 ds.send(dp);
                 Thread.sleep(50);
                 ds.send(dp);
-            } catch (Exception ex) {
-                e = ex;
+            } catch (Exception e) {
+                log.println("ClassicWriter [" + ds.getLocalAddress() + "]");
+                throw new RuntimeException("ClassicWriter threw exception: " + e);
+            } finally {
+                log.println("ClassicWriter finished");
             }
         }
+
+        @Override
+        public void close() throws IOException {
+            ds.close();
+        }
     }
 
-    public static class NioWriter implements Sprintable {
+    public static class NioWriter implements Runnable, AutoCloseable {
+        final DatagramChannel dc;
         final int port;
-        Exception e = null;
 
-        NioWriter(int port) {
+        NioWriter(int port) throws IOException {
+            this.dc = DatagramChannel.open();
             this.port = port;
         }
 
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
-        }
-
         public void run() {
             try {
                 DatagramChannel dc = DatagramChannel.open();
                 ByteBuffer bb = ByteBuffer.allocateDirect(256);
                 bb.put("hello".getBytes());
                 bb.flip();
-                InetAddress address = InetAddress.getLocalHost();
+                InetAddress address = InetAddress.getLoopbackAddress();
                 InetSocketAddress isa = new InetSocketAddress(address, port);
                 dc.send(bb, isa);
                 Thread.sleep(50);
                 dc.send(bb, isa);
             } catch (Exception ex) {
-                e = ex;
+                log.println("ClassicWriter [" + dc.socket().getLocalAddress() + "]");
```

Comment:
This is not the `ClassicWriter` but the `NioWriter`

Reply:
Same is true for the other classes, a fix was not included in the previous commit and will be in the next one